pkg/cmd/gist/edit/edit.go

```diff
@@ -99,89 +111,152 @@ func editRun(opts *EditOptions) error {
 
 	filesToUpdate := map[string]string{}
 
-	for {
-		filename := opts.Filename
-		candidates := []string{}
-		for filename := range gist.Files {
-			candidates = append(candidates, filename)
+	addFilename := opts.AddFilename
+	cs := opts.IO.ColorScheme()
+
+	if addFilename != "" {
+		//Add files to an existing gist.
+		filenamePath, fileExists, err := processFiles(addFilename)
+		if err != nil {
+			return fmt.Errorf("%s %s", cs.Red("!"), err)
 		}
 
-		sort.Strings(candidates)
+		filesToAdd := map[string]*shared.GistFile{}
 
-		if filename == "" {
-			if len(candidates) == 1 {
-				filename = candidates[0]
-			} else {
-				if !opts.IO.CanPrompt() {
-					return errors.New("unsure what file to edit; either specify --filename or run interactively")
-				}
-				err = prompt.SurveyAskOne(&survey.Select{
-					Message: "Edit which file?",
-					Options: candidates,
-				}, &filename)
+		filename := filepath.Base(filenamePath)
 
-				if err != nil {
-					return fmt.Errorf("could not prompt: %w", err)
-				}
+		if fileExists {
+			content, err := ioutil.ReadFile(filenamePath)
+			if err != nil {
+				return fmt.Errorf("%s failed to read file %s: %w", cs.FailureIcon(), addFilename, err)
 			}
-		}
 
-		if _, ok := gist.Files[filename]; !ok {
-			return fmt.Errorf("gist has no file %q", filename)
-		}
+			if string(content) == "" {
+				return fmt.Errorf("%s Contents can't be empty", cs.FailureIcon())
+			}
 
-		editorCommand, err := cmdutil.DetermineEditor(opts.Config)
-		if err != nil {
-			return err
+			filesToAdd[filename] = &shared.GistFile{
+				Filename: filename,
+				Content:  string(content),
+			}
+			gist.Files = filesToAdd
+		} else {
+			editorCommand, err := cmdutil.DetermineEditor(opts.Config)
+			if err != nil {
+				return err
+			}
+
+			text, err := opts.Add(editorCommand, filename, opts.IO)
+			if err != nil {
+				return err
+			}
+
+			if text == "" {
+				return fmt.Errorf("%s Contents can't be empty", cs.Red("!"))
+			}
+
+			filesToAdd[filename] = &shared.GistFile{
+				Filename: filename,
+				Content:  text,
+			}
+
+			gist.Files = filesToAdd
 		}
-		text, err := opts.Edit(editorCommand, filename, gist.Files[filename].Content, opts.IO)
 
+		err = updateGist(apiClient, ghinstance.OverridableDefault(), gist)
 		if err != nil {
 			return err
 		}
 
-		if text != gist.Files[filename].Content {
-			gistFile := gist.Files[filename]
-			gistFile.Content = text // so it appears if they re-edit
-			filesToUpdate[filename] = text
-		}
+		completionMessage := filename + " added to gist"
 
-		if !opts.IO.CanPrompt() {
-			break
-		}
+		fmt.Fprintf(opts.IO.Out, "%s %s\n", cs.SuccessIconWithColor(cs.Green), completionMessage)
 
-		if len(candidates) == 1 {
-			break
-		}
+	} else {
```

Comment:
this conditional is getting out of hand. I'd like to see helper functions and early exits for adding and editing files.